lib/fileset/internal.nix

```diff
@@ -203,22 +210,22 @@ rec {
       // value;
 
   /*
-    Simplify a filesetTree recursively:
-    - Replace all directories that have no files with `null`
+    A normalisation of a filesetTree suitable for use in builtins.path-based filtering:
+    - Replace all directories that have no files with `null`.
       This removes directories that would be empty
-    - Replace all directories with all files with `"directory"`
+    - Replace all directories with all files with `"directory"`.
       This speeds up the source filter function
 
     Note that this function is strict, it evaluates the entire tree
 
     Type: Path -> filesetTree -> filesetTree
   */
-  _simplifyTree = path: tree:
+  _normaliseTreeFilter = path: tree:
```

Comment:
Nice! Performance is really the main reason though, running `./benchmark HEAD` with this change:

```
Benchmarking expression toSource { root = ./.; fileset = ./.; }
Mean CPU time 0.070476 (σ = 0.0035255) for 10 runs is 103.50068950527445% (σ = 6.396296936915695%) of the old value 0.0680923 (σ = 0.00247089)
Statistic .envs.elements (313193) is 160.6085% (+118189) of the old value 195004
Statistic .envs.number (219488) is 116.5091% (+31101) of the old value 188387
Statistic .gc.totalBytes (21491904) is 108.0092% (+1593696) of the old value 19898208
Statistic .nrFunctionCalls (194560) is 119.0260% (+31100) of the old value 163460
Statistic .nrPrimOpCalls (157243) is 100.0012% (+2) of the old value 157241
Statistic .nrThunks (123912) is 100.0064% (+8) of the old value 123904
Statistic .sets.elements (26581) is 100.0338% (+9) of the old value 26572
Statistic .sets.number (3672) is 100.0272% (+1) of the old value 3671
Statistic .symbols.number (984) is 100.9230% (+9) of the old value 975
Statistic .values.number (264327) is 100.0034% (+9) of the old value 264318
10 stats differ between the current tree and HEAD
```

It's not _that_ significant tbh, but I wonder where to draw the line between optimisations for the sake of code clarity.

Reply:
I also have an idea to do a transformation with `mapAttrs'` that could be used with a much simpler `foldl'` tracing: Carry some additional context with each node to denote depth, and add indentation to the keys directly. Then the trace wrapper has only three branches and needs no helpers:
- `null`: skip
- string value: trace value
- attrs: trace name, recurse

Reply:
Oh I was just checking the performance again, but was confused because it changed. Until I realised that you edited the code :). Though it's less efficient now:

```
Benchmarking expression toSource { root = ./.; fileset = ./.; }
Mean CPU time 0.109669 (σ = 0.00826011) for 10 runs is 104.26395649528446% (σ = 10.231057763965339%) of the old value 0.105184 (σ = 0.00661567)
Statistic .envs.elements (364517) is 186.9279% (+169513) of the old value 195004
Statistic .envs.number (239705) is 127.2407% (+51318) of the old value 188387
Statistic .gc.totalBytes (24858208) is 124.9268% (+4960000) of the old value 19898208
Statistic .list.elements (113956) is 110.5875% (+10910) of the old value 103046
Statistic .nrFunctionCalls (205446) is 125.6857% (+41986) of the old value 163460
Statistic .nrLookups (57647) is 263.4809% (+35768) of the old value 21879
Statistic .nrPrimOpCalls (158799) is 100.9908% (+1558) of the old value 157241
Statistic .nrThunks (170571) is 137.6638% (+46667) of the old value 123904
Statistic .sets.elements (63906) is 240.5012% (+37334) of the old value 26572
Statistic .sets.number (16112) is 438.8994% (+12441) of the old value 3671
Statistic .symbols.number (989) is 101.4358% (+14) of the old value 975
Statistic .values.number (310985) is 117.6556% (+46667) of the old value 264318
12 stats differ between the current tree and HEAD
```

25% more memory usage. I think that's too much just for cleaner code.

Additionally, in case https://github.com/NixOS/nix/issues/8820 gets implemented, the `_normaliseTreeFilter` function wouldn't be necessary anymore, at which point we wouldn't need this internal abstraction either.